{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Informatics code for Younginger, Stewart, Balkan and Ballhorn\n",
    "## \"Stable coexistence or competitive exclusion? \n",
    "## Fern endophytes shift our understanding of microbial temporal turnover\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Software needed for analysis:\n",
    "FastQC v0.11.5\n",
    "usearch v10.0.247M_i86linux64\n",
    "jun_jul_renamer.sh\n",
    "control_renamer.sh\n",
    "R1_renamer.sh\n",
    "\n",
    "## Sample processing\n",
    "\n",
    "### Concatenate the R1 reads since they were demultiplexed:\n",
    "\n",
    "cat *_R1_* > ../FastQC/jun_jul_R1.fastq.gz\n",
    "\n",
    "##### Total raw R1 reads for each month #####\n",
    "\n",
    "- Jun_Jul: 2605987\n",
    "- Aug_Sep: 2440660\n",
    "- Oct_Nov: 3310806\n",
    "- Dec_Jan: 2465935\n",
    "- preprocessed Apr_May: 15566066\n",
    "- demultiplexed Apr_May: 7955646\n",
    "\n",
    "## Demultiplex April May reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_demux Undetermined_S0_L001_R1_001.fastq -reverse Undetermined_S0_L001_R2_001.fastq -index Undetermined_S0_L001_I1_001.fastq -barcodes barcodes_6.fa -fastqout apr_may_R1.fastq -output2 apr_may_R2.fastq\n",
    "\n",
    "\t03:30 110Mb   100.0% Demuxed 7955646 / 15566066 (51.1%)\n",
    "\n",
    "## Consider using only R1 reads and also merging, running the analyses side-by-side to see how the results differ...\n",
    "\t## Will merge first though\n",
    "\n",
    "## Also, need to look at how the files are labeled and rename the control files to match the appropriate formatting\n",
    "\t## Jun_Jul will need to be renamed to match the other months, too!\n",
    "\n",
    "# Changed the Jun_Jul samples with the following script:\n",
    "\n",
    "jun_jul_renamer.sh\n",
    "\n",
    "# Will now change up the names of the controls in the remaining samples\n",
    "\n",
    "control_renamer.sh\n",
    "\n",
    "## Merging reads\n",
    "\n",
    "## Apr_May\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs apr_may_R1.fastq -fastqout ../Merged_reads/apr_may.fastq -fastq_maxdiffs 300\n",
    "\n",
    "\n",
    "Merging\n",
    "  Fwd apr_may_R1.fastq\n",
    "  Rev apr_may_R2.fastq\n",
    "  Keep read labels\n",
    "\n",
    "00:01 71Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "01:14 751Mb   100.0% 91.4% merged\n",
    "\n",
    "Totals:\n",
    "   7955646  Pairs (8.0M)\n",
    "   7274761  Merged (7.3M, 91.44%)\n",
    "   1423348  Alignments with zero diffs (17.89%)\n",
    "    641535  Too many diffs (> 300) (8.06%)\n",
    "     39350  No alignment found (0.49%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   7469521  Staggered pairs (93.89%) merged & trimmed\n",
    "    205.78  Mean alignment length\n",
    "    209.40  Mean merged length\n",
    "      0.33  Mean fwd expected errors\n",
    "      1.90  Mean rev expected errors\n",
    "      0.14  Mean merged expected errors\n",
    "      \n",
    "## Jun_Jul\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/jun_jul.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:18 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:18 765Mb   100.0% 91.9% merged\n",
    "\n",
    "Totals:\n",
    "   2605987  Pairs (2.6M)\n",
    "   2395636  Merged (2.4M, 91.93%)\n",
    "    193274  Alignments with zero diffs (7.42%)\n",
    "    175297  Too many diffs (> 300) (6.73%)\n",
    "     35054  No alignment found (1.35%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2344067  Staggered pairs (89.95%) merged & trimmed\n",
    "    199.25  Mean alignment length\n",
    "    216.07  Mean merged length\n",
    "      0.64  Mean fwd expected errors\n",
    "      1.70  Mean rev expected errors\n",
    "      0.19  Mean merged expected errors\n",
    "\n",
    "#### Wondering if I need to deal with differences in header lines between Apr_May and the rest of the files:\n",
    "\n",
    "@M02149:71:000000000-AAMFM:1:1101:13889:1730;sample=BY-12-1-5;\n",
    "\n",
    "@BY-10-1-6.2\n",
    "\n",
    "## Will try to address it with the following argument: -relabel @\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs apr_may_R1.fastq -fastqout ../Merged_reads/apr_may_V2.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\t# Nope, that didn't do it. Hopefully it will get taken care of later in processing  \n",
    "    \n",
    "## Aug_Sep\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/aug_sep.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:15 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:15 765Mb   100.0% 54.8% merged\n",
    "\n",
    "Totals:\n",
    "   2440660  Pairs (2.4M)\n",
    "   1337858  Merged (1.3M, 54.82%)\n",
    "      9220  Alignments with zero diffs (0.38%)\n",
    "   1083301  Too many diffs (> 300) (44.39%)\n",
    "         4  Fwd tails Q <= 2 trimmed (0.00%)\n",
    "        41  Rev tails Q <= 2 trimmed (0.00%)\n",
    "        87  Fwd too short (< 64) after tail trimming (0.00%)\n",
    "       309  Rev too short (< 64) after tail trimming (0.01%)\n",
    "     19105  No alignment found (0.78%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2306121  Staggered pairs (94.49%) merged & trimmed\n",
    "    201.15  Mean alignment length\n",
    "    210.86  Mean merged length\n",
    "      1.16  Mean fwd expected errors\n",
    "      3.87  Mean rev expected errors\n",
    "      0.62  Mean merged expected errors\n",
    "\n",
    "# Too many diffs resulted in 44.39% of the reads getting discarded! Did this happen before? \n",
    "\t# May need to just continue with the R1s...\n",
    "\t# But what if I run it with 400 max diffs? \n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/aug_sep_V2.fastq -relabel @ -fastq_maxdiffs 400\n",
    "\t# Results in the exact same output. Really weird and an indication that I should be using the R1 reads only\n",
    "\n",
    "## Oct_Nov\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout oct_nov.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:21 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:21 765Mb   100.0% 90.4% merged\n",
    "\n",
    "Totals:\n",
    "   3310806  Pairs (3.3M)\n",
    "   2992916  Merged (3.0M, 90.40%)\n",
    "     34999  Alignments with zero diffs (1.06%)\n",
    "    306191  Too many diffs (> 300) (9.25%)\n",
    "     11699  No alignment found (0.35%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2946717  Staggered pairs (89.00%) merged & trimmed\n",
    "    197.32  Mean alignment length\n",
    "    219.33  Mean merged length\n",
    "      0.64  Mean fwd expected errors\n",
    "      2.59  Mean rev expected errors\n",
    "      0.58  Mean merged expected errors\n",
    "\n",
    "## Dec_Jan\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/dec_jan.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:16 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:16 765Mb   100.0% 94.2% merged\n",
    "\n",
    "Totals:\n",
    "   2465935  Pairs (2.5M)\n",
    "   2323617  Merged (2.3M, 94.23%)\n",
    "     63861  Alignments with zero diffs (2.59%)\n",
    "    133325  Too many diffs (> 300) (5.41%)\n",
    "      8993  No alignment found (0.36%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2158636  Staggered pairs (87.54%) merged & trimmed\n",
    "    196.24  Mean alignment length\n",
    "    221.91  Mean merged length\n",
    "      0.80  Mean fwd expected errors\n",
    "      2.38  Mean rev expected errors\n",
    "      0.33  Mean merged expected errors\n",
    "\n",
    "## Put them all together\n",
    "\n",
    "cat *.fastq > temporal_merged.fastq\n",
    "\n",
    "###########\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix temporal_merged.fastq -output phix_filtered.fastq -alnout phix_hits.txt\n",
    "\n",
    "00:00 277Mb   100.0% Word stats\n",
    "00:00 277Mb   100.0% Alloc rows\n",
    "00:00 277Mb   100.0% Build index\n",
    "00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "01:08 933Mb   100.0% Filtering for phix, 838106 hits (5.1%)\n",
    "\n",
    "cp phix_filtered.fastq ../Redux_processing/\n",
    "\n",
    "mv phix_filtered.fastq merged_phix_filtered.fastq\n",
    "\n",
    "# Okay, will move on to R now...\n",
    "\n",
    "# Well, that's disappointing: only have the apr_may samples in the otu table. Will check to see if this is the case in the zotu tables\n",
    "\n",
    "# The reads are in the merged files, but I suspect the difference in the header lines is throwing off the commands. \n",
    "\t# Recall needing to reformat the header lines for the apr_may samples, but will have to look through the old code to \n",
    "\t# figure out what I did\n",
    "\t\t# I did use a script fastq_header_changer3.py to change the apr_may files previously, but maybe I can get around it with\n",
    "\t\t# the -sample_delim . option for the otu_tab command, if all the reads are present up until that point...\n",
    "\n",
    "fastq_header_changer3.py apr_may.fastq > apr_may_V2.fastq\n",
    "\n",
    "# A quick check of the headers for comparison:\n",
    "\n",
    "less apr_may_V2.fastq: @BY-12-1-5.1101138891730\n",
    "less aug_sep.fastq: @BY-9-4-9.1337858\n",
    "\n",
    "fastq_header_changer3.py merged_phix_filtered.fastq > merged_phix_filtered_V2.fastq\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered.fastq\n",
    "\t15487871\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered_V2.fastq \n",
    "\t15487871\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter merged_phix_filtered_V2.fastq -fastq_maxee 1.0 -fastaout merged_filtered_V2.fasta\n",
    "\n",
    "01:44 683Mb   100.0% Filtering, 95.6% passed\n",
    "  15486682  Reads (15.5M)                   \n",
    "    677084  Discarded reads with expected errs > 1.00\n",
    "  14809598  Filtered reads (14.8M, 95.6%)\n",
    "    \n",
    "usearch10.0.274M_i86linux64 -orient merged_filtered_V2.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout merged_oriented_V2.fasta -tabbedout merge_orient_V2.txt\n",
    "\n",
    "01:40 1.0Gb   100.0% 14678970 plus (99.1%), 985 minus (0.0%), 129643 undet. (0.9%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques merged_oriented_V2.fasta -fastaout merged_uniques_V2.fasta -sizeout -relabel Uniq\n",
    "\n",
    "01:02 7.6Gb  14679955 seqs, 306010 uniques, 199184 singletons (65.1%)\n",
    "01:02 7.6Gb  Min size 1, median 1, max 9957620, avg 47.97\n",
    "01:04 5.3Gb   100.0% Writing merged_uniques_V2.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 merged_uniques_V2.fasta -zotus merged_zotus_V2.fasta\n",
    "\n",
    "00:01 130Mb   100.0% Reading merged_uniques_V2.fasta\n",
    "00:02 120Mb   100.0% 994 amplicons, 2592535 bad (size >= 8)  \n",
    "00:16 131Mb   100.0% 987 good, 7 chimeras                  \n",
    "00:16 131Mb   100.0% Writing zotus\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab merged_phix_filtered_V2.fastq -otus merged_zotus_V2.fasta -otutabout merged_zotutab_V3.txt -mapout merged_zmap_V3.txt -sample_delim . \n",
    "\n",
    "42:18 726Mb   100.0% Searching, 99.6% matched\n",
    "15423973 / 15486682 mapped to OTUs (99.6%)\n",
    "42:18 726Mb  Writing merged_zotutab_V3.txt\n",
    "42:18 726Mb  Writing merged_zotutab_V3.txt ...done.\n",
    "\n",
    "\n",
    "/workspace/scratch/obrett/RDP\n",
    "\n",
    "wget https://drive5.com/sintax/rdp_its_v2.fa.gz\n",
    "\n",
    "gunzip rdp_its_v2.fa.gz\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax rdp_its_v2.fa -output rdp_its_v2.udb\n",
    "\n",
    "## This output has a taxonomic assignment for almost every rank which is helpful. Should compare between the two further. \n",
    "\n",
    "\t# Think I'm going to work with the RDP...\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus_V2.fasta -db ../../../RDP/rdp_its_v2.udb -strand both -tabbedout merged_zotus_V2_rdp.sintax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Processing for within plant\n",
    "\n",
    "## First need to rename the fastq files (and potentially the headers) to make sure they process neatly downstream\n",
    "\n",
    "## Will modify the shell script jun_jul_renamer.sh for the within-plant\n",
    "\n",
    "within_plant_renamer.sh\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/within_p_merged2.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "Totals:\n",
    "    675853  Pairs (675.9k)\n",
    "    583714  Merged (583.7k, 86.37%)\n",
    "    229030  Alignments with zero diffs (33.89%)\n",
    "     74970  Too many diffs (> 300) (11.09%)\n",
    "     17169  No alignment found (2.54%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "    581026  Staggered pairs (85.97%) merged & trimmed\n",
    "    203.81  Mean alignment length\n",
    "    219.84  Mean merged length\n",
    "      0.37  Mean fwd expected errors\n",
    "      1.73  Mean rev expected errors\n",
    "      0.17  Mean merged expected errors\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix within_p_merged2.fastq -output within_p_phix_f.fastq -alnout phix_hits.txt\n",
    "\n",
    "00:02 935Mb   100.0% Filtering for phix, 12380 hits (2.1%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter within_p_phix_f.fastq -fastq_maxee 1.0 -fastaout within_p_filtered.fasta\n",
    "\n",
    "00:04 683Mb   100.0% Filtering, 96.6% passed\n",
    "    571334  Reads (571.3k)                  \n",
    "     19582  Discarded reads with expected errs > 1.00\n",
    "    551752  Filtered reads (551.8k, 96.6%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient within_p_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout within_p_oriented.fasta -tabbedout within_p_oriented.txt\n",
    "\n",
    "00:04 1.0Gb   100.0% 548048 plus (99.3%), 2 minus (0.0%), 3702 undet. (0.7%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques within_p_oriented.fasta -fastaout within_p_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:02 898Mb  548050 seqs, 20995 uniques, 15276 singletons (72.8%)\n",
    "00:02 898Mb  Min size 1, median 1, max 180839, avg 26.10\n",
    "00:03 898Mb   100.0% Writing within_p_uniques.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 within_p_uniques.fasta -zotus within_p_zotus.fasta\n",
    "\n",
    "00:00 49Mb    100.0% Reading within_p_uniques.fasta\n",
    "00:00 30Mb    100.0% 620 amplicons, 38994 bad (size >= 8)\n",
    "00:05 41Mb    100.0% 615 good, 5 chimeras                \n",
    "00:05 41Mb    100.0% Writing zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab within_p_phix_f.fastq -otus within_p_zotus.fasta -otutabout within_p_zotutab.txt -mapout within_p_zmap.txt\n",
    "\n",
    "03:32 726Mb   100.0% Searching, 99.1% matchedchedg, 98.6% matched\n",
    "566207 / 571334 mapped to OTUs (99.1%)\n",
    "03:32 726Mb  Writing within_p_zotutab.txt\n",
    "03:32 726Mb  Writing within_p_zotutab.txt ...done.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax within_p_zotus.fasta -db ../../../RDP/rdp_its_v2.udb -tabbedout within_p_zotus.sintax -strand both\n",
    "\n",
    "## Done. Will incorporate into R for future analyses\n",
    "\n",
    "# Just copied over the following files to the following directory for the # # within plant analysis:\n",
    "\n",
    "/Users/brett/Documents/Temporal_turnover/Informatics\n",
    "within_p_zotus.sintax\n",
    "within_p_zotutab.txt\n",
    "\n",
    "# Need to make an updated RDP sintax file for the zotu table:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus_V2.fasta -db ../../../RDP/rdp_its_v2.udb -strand both -tabbedout merged_zotus_V2_rdp.sintax\n",
    "\n",
    "# Need the following files for zotu analysis:\n",
    "- merged_zotus_V2.fasta\n",
    "- merged_zotus_V2_rdp.sintax\n",
    "- merged_zotutab_V3.txt\n",
    "- map_file16.txt\n",
    "\n",
    "# Need to reformat merged_zotus_V2_rdp.sintax so it matches the formatting of merged_otus_V3.sintax\n",
    "\t# Saved as merged_zotus_V3_rdp.sintax\n",
    "\n",
    "# Consider removing conf estimates and incertae sedis from tax table\n",
    "\n",
    "# Need to re-run line 1629 from above with the -sample_delim . argument\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Within_plant/Merged_reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab within_p_phix_f.fastq -otus within_p_zotus.fasta -otutabout within_p_zotutab2.txt -mapout within_p_zmap2.txt -sample_delim .\n",
    "\n",
    "within_p_zotus.sintax\n",
    "within_p_zotutab2.txt\n",
    "within_p_map_file.txt\n",
    "\n",
    "# Plotted an ordination and taxonomy plot for the within-p\n",
    "\t# See that there are multiples of the same genus/family on the \n",
    "\t# taxonomy plot. Need to delete these conf. estimates since it makes \n",
    "\t# each genus seem distinct. Will also delete the f: and g: designations\n",
    "\t\n",
    "# Will update this on both taxonomy files for the big dataset and the within-p\n",
    "\n",
    "Originals:\n",
    "- 'within_p_zotus2.sintax.txt'\n",
    "- 'merged_zotus_V3_rdp.sintax'\n",
    "\n",
    "Edited:\n",
    "- 'within_p_zotus3.sintax'\n",
    "- 'merged_zotus_V4_rdp.sintax'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
