{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Informatics code for \"Stable coexistence or competitive exclusion? \n",
    "# Fern endophytes shift our understanding of microbial temporal turnover\"\n",
    "## Younginger, Stewart, Balkan and Ballhorn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Software needed for analysis:\n",
    "FastQC v0.11.5\n",
    "usearch v10.0.247M_i86linux64\n",
    "jun_jul_renamer.sh\n",
    "control_renamer.sh\n",
    "R1_renamer.sh\n",
    "within_plant_renamer.sh\n",
    "\n",
    "# Sample processing\n",
    "\n",
    "# Concatenate the R1 reads since they were demultiplexed:\n",
    "\n",
    "cat *_R1_* > ../FastQC/jun_jul_R1.fastq.gz # Same command for all demultiplexed runs\n",
    "\n",
    "##### Total raw R1 reads for each month #####\n",
    "\n",
    "- Jun_Jul: 2605987\n",
    "- Aug_Sep: 2440660\n",
    "- Oct_Nov: 3310806\n",
    "- Dec_Jan: 2465935\n",
    "- preprocessed Apr_May: 15566066\n",
    "- demultiplexed Apr_May: 7955646\n",
    "\n",
    "# Demultiplex April May reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_demux Undetermined_S0_L001_R1_001.fastq -reverse \\\n",
    "    Undetermined_S0_L001_R2_001.fastq -index Undetermined_S0_L001_I1_001.fastq -barcodes barcodes_6.fa -fastqout \\\n",
    "    apr_may_R1.fastq -output2 apr_may_R2.fastq\n",
    "\n",
    "#\t03:30 110Mb   100.0% Demuxed 7955646 / 15566066 (51.1%)\n",
    "\n",
    "# Change the Jun_Jul filenames to SampleIDs with the following script:\n",
    "\n",
    "jun_jul_renamer.sh\n",
    "\n",
    "# Change the names of the controls in each sequencing run\n",
    "\n",
    "control_renamer.sh\n",
    "\n",
    "# Merge forward and reverse reads\n",
    "\n",
    "# Apr_May\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs apr_may_R1.fastq -fastqout ../Merged_reads/apr_may.fastq \\\n",
    "    -fastq_maxdiffs 300\n",
    "\n",
    "\n",
    "# Merging\n",
    "#   Fwd apr_may_R1.fastq\n",
    "#   Rev apr_may_R2.fastq\n",
    "#   Keep read labels\n",
    "\n",
    "# 00:01 71Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "# 01:14 751Mb   100.0% 91.4% merged\n",
    "\n",
    "# Totals:\n",
    "#    7955646  Pairs (8.0M)\n",
    "#    7274761  Merged (7.3M, 91.44%)\n",
    "#    1423348  Alignments with zero diffs (17.89%)\n",
    "#     641535  Too many diffs (> 300) (8.06%)\n",
    "#      39350  No alignment found (0.49%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#    7469521  Staggered pairs (93.89%) merged & trimmed\n",
    "#     205.78  Mean alignment length\n",
    "#     209.40  Mean merged length\n",
    "#       0.33  Mean fwd expected errors\n",
    "#       1.90  Mean rev expected errors\n",
    "#       0.14  Mean merged expected errors\n",
    "      \n",
    "# Jun_Jul\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/jun_jul.fastq -relabel \\\n",
    "    @ -fastq_maxdiffs 300\n",
    "\n",
    "# 00:18 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "# 00:18 765Mb   100.0% 91.9% merged\n",
    "\n",
    "# Totals:\n",
    "#    2605987  Pairs (2.6M)\n",
    "#    2395636  Merged (2.4M, 91.93%)\n",
    "#     193274  Alignments with zero diffs (7.42%)\n",
    "#     175297  Too many diffs (> 300) (6.73%)\n",
    "#      35054  No alignment found (1.35%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#    2344067  Staggered pairs (89.95%) merged & trimmed\n",
    "#     199.25  Mean alignment length\n",
    "#     216.07  Mean merged length\n",
    "#       0.64  Mean fwd expected errors\n",
    "#       1.70  Mean rev expected errors\n",
    "#       0.19  Mean merged expected errors\n",
    "    \n",
    "# Aug_Sep\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/aug_sep.fastq -relabel @ \\\n",
    "    -fastq_maxdiffs 300\n",
    "\n",
    "# 00:15 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "# 00:15 765Mb   100.0% 54.8% merged\n",
    "\n",
    "# Totals:\n",
    "#    2440660  Pairs (2.4M)\n",
    "#    1337858  Merged (1.3M, 54.82%)\n",
    "#       9220  Alignments with zero diffs (0.38%)\n",
    "#    1083301  Too many diffs (> 300) (44.39%)\n",
    "#          4  Fwd tails Q <= 2 trimmed (0.00%)\n",
    "#         41  Rev tails Q <= 2 trimmed (0.00%)\n",
    "#         87  Fwd too short (< 64) after tail trimming (0.00%)\n",
    "#        309  Rev too short (< 64) after tail trimming (0.01%)\n",
    "#      19105  No alignment found (0.78%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#    2306121  Staggered pairs (94.49%) merged & trimmed\n",
    "#     201.15  Mean alignment length\n",
    "#     210.86  Mean merged length\n",
    "#       1.16  Mean fwd expected errors\n",
    "#       3.87  Mean rev expected errors\n",
    "#       0.62  Mean merged expected errors\n",
    "\n",
    "# Oct_Nov\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout oct_nov.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "# 00:21 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "# 00:21 765Mb   100.0% 90.4% merged\n",
    "\n",
    "# Totals:\n",
    "#    3310806  Pairs (3.3M)\n",
    "#    2992916  Merged (3.0M, 90.40%)\n",
    "#      34999  Alignments with zero diffs (1.06%)\n",
    "#     306191  Too many diffs (> 300) (9.25%)\n",
    "#      11699  No alignment found (0.35%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#    2946717  Staggered pairs (89.00%) merged & trimmed\n",
    "#     197.32  Mean alignment length\n",
    "#     219.33  Mean merged length\n",
    "#       0.64  Mean fwd expected errors\n",
    "#       2.59  Mean rev expected errors\n",
    "#       0.58  Mean merged expected errors\n",
    "\n",
    "# Dec_Jan\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/dec_jan.fastq -relabel @ \\\n",
    "    -fastq_maxdiffs 300\n",
    "\n",
    "# 00:16 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "# 00:16 765Mb   100.0% 94.2% merged\n",
    "\n",
    "# Totals:\n",
    "#    2465935  Pairs (2.5M)\n",
    "#    2323617  Merged (2.3M, 94.23%)\n",
    "#      63861  Alignments with zero diffs (2.59%)\n",
    "#     133325  Too many diffs (> 300) (5.41%)\n",
    "#       8993  No alignment found (0.36%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#    2158636  Staggered pairs (87.54%) merged & trimmed\n",
    "#     196.24  Mean alignment length\n",
    "#     221.91  Mean merged length\n",
    "#       0.80  Mean fwd expected errors\n",
    "#       2.38  Mean rev expected errors\n",
    "#       0.33  Mean merged expected errors\n",
    "\n",
    "# Combine all merged reads\n",
    "\n",
    "cat *.fastq > temporal_merged.fastq\n",
    "\n",
    "# Filter out phix sequences from the reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix temporal_merged.fastq -output phix_filtered.fastq -alnout phix_hits.txt\n",
    "\n",
    "# 00:00 277Mb   100.0% Word stats\n",
    "# 00:00 277Mb   100.0% Alloc rows\n",
    "# 00:00 277Mb   100.0% Build index\n",
    "# 00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "# 01:08 933Mb   100.0% Filtering for phix, 838106 hits (5.1%)\n",
    "\n",
    "mv phix_filtered.fastq merged_phix_filtered.fastq\n",
    "\n",
    "# Change head headers of the Apr_May sequencing run\n",
    "\n",
    "fastq_header_changer3.py merged_phix_filtered.fastq > merged_phix_filtered_V2.fastq\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered.fastq\n",
    "# \t15487871\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered_V2.fastq \n",
    "# \t15487871\n",
    "\n",
    "# Filter out reads with expected number of errors greater than 1.0: \\\n",
    "#     https://www.drive5.com/usearch/manual/exp_errs.html\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter merged_phix_filtered_V2.fastq -fastq_maxee 1.0 -fastaout \\\n",
    "    merged_filtered_V2.fasta\n",
    "\n",
    "# 01:44 683Mb   100.0% Filtering, 95.6% passed\n",
    "#   15486682  Reads (15.5M)                   \n",
    "#     677084  Discarded reads with expected errs > 1.00\n",
    "#   14809598  Filtered reads (14.8M, 95.6%)\n",
    "\n",
    "# Re-orient the reads to forward, filter out reads with an undetermined orientation\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient merged_filtered_V2.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout \\\n",
    "    merged_oriented_V2.fasta -tabbedout merge_orient_V2.txt\n",
    "\n",
    "# 01:40 1.0Gb   100.0% 14678970 plus (99.1%), 985 minus (0.0%), 129643 undet. (0.9%)\n",
    "\n",
    "# Find unique sequences for denoising\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques merged_oriented_V2.fasta -fastaout merged_uniques_V2.fasta -sizeout \\\n",
    "    -relabel Uniq\n",
    "\n",
    "# 01:02 7.6Gb  14679955 seqs, 306010 uniques, 199184 singletons (65.1%)\n",
    "# 01:02 7.6Gb  Min size 1, median 1, max 9957620, avg 47.97\n",
    "# 01:04 5.3Gb   100.0% Writing merged_uniques_V2.fasta\n",
    "\n",
    "# De-noise reads into zero-radius otus (i.e. ESVs or ASVs)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 merged_uniques_V2.fasta -zotus merged_zotus_V2.fasta\n",
    "\n",
    "# 00:01 130Mb   100.0% Reading merged_uniques_V2.fasta\n",
    "# 00:02 120Mb   100.0% 994 amplicons, 2592535 bad (size >= 8)  \n",
    "# 00:16 131Mb   100.0% 987 good, 7 chimeras                  \n",
    "# 00:16 131Mb   100.0% Writing zotus\n",
    "\n",
    "# Make otu table with original sequencing reads mapped to zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab merged_phix_filtered_V2.fastq -otus merged_zotus_V2.fasta -otutabout \\\n",
    "    merged_zotutab_V3.txt -mapout merged_zmap_V3.txt -sample_delim . \n",
    "\n",
    "# 42:18 726Mb   100.0% Searching, 99.6% matched\n",
    "# 15423973 / 15486682 mapped to OTUs (99.6%)\n",
    "# 42:18 726Mb  Writing merged_zotutab_V3.txt\n",
    "# 42:18 726Mb  Writing merged_zotutab_V3.txt ...done.\n",
    "\n",
    "# Download the most recent RDP classifier (V2) for taxonomy\n",
    "\n",
    "wget https://drive5.com/sintax/rdp_its_v2.fa.gz\n",
    "\n",
    "gunzip rdp_its_v2.fa.gz\n",
    "\n",
    "# Make taxonomy database\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax rdp_its_v2.fa -output rdp_its_v2.udb\n",
    "\n",
    "# Assign taxonomy to the zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus_V2.fasta -db ../../../RDP/rdp_its_v2.udb -strand both -tabbedout \\\n",
    "    merged_zotus_V2_rdp.sintax\n",
    "\n",
    "# Output files for analyses in R:\n",
    "\n",
    "# 'merged_zotutab_V3.txt'\n",
    "# 'map_file16.txt'\n",
    "# 'merged_zotus_V4_rdp.sintax' \n",
    "#     Note: For the sintax file, it was modified in a text editor from the following format \\\n",
    "#     (merged_zotus_V2_rdp.sintax)\n",
    "\n",
    "#     Zotu2\td:Fungi(1.0000),p:Ascomycota(1.0000),c:Leotiomycetes(0.7800),o:Helotiales(0.7800),\\\n",
    "#                     f:Helotiaceae(0.7700),g:Hymenoscyphus(0.6200),s:Articulospora_tetracladia(0.6200)\t+\td:\\\n",
    "#                                     Fungi,p:Ascomycota\n",
    "\n",
    "#     To the format below (merged_zotus_V4_rdp.sintax)\n",
    "\n",
    "#     Zotu2\tFungi\tAscomycota\tLeotiomycetes\tHelotiales\tHelotiaceae\tHymenoscyphus\tArticulospora_tetracladia\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Processing for within-plant analysis\n",
    "\n",
    "# Rename the fastq files (and the headers) to make sure they process neatly downstream\n",
    "\n",
    "witin_plant_renamer.sh\n",
    "\n",
    "# Merge forward and reverse reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/within_p_merged2.fastq \\\n",
    "    -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "# Totals:\n",
    "#     675853  Pairs (675.9k)\n",
    "#     583714  Merged (583.7k, 86.37%)\n",
    "#     229030  Alignments with zero diffs (33.89%)\n",
    "#      74970  Too many diffs (> 300) (11.09%)\n",
    "#      17169  No alignment found (2.54%)\n",
    "#          0  Alignment too short (< 16) (0.00%)\n",
    "#     581026  Staggered pairs (85.97%) merged & trimmed\n",
    "#     203.81  Mean alignment length\n",
    "#     219.84  Mean merged length\n",
    "#       0.37  Mean fwd expected errors\n",
    "#       1.73  Mean rev expected errors\n",
    "#       0.17  Mean merged expected errors\n",
    "\n",
    "# Filter out phix sequences from the reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix within_p_merged2.fastq -output within_p_phix_f.fastq -alnout \\\n",
    "    phix_hits.txt\n",
    "\n",
    "# 00:02 935Mb   100.0% Filtering for phix, 12380 hits (2.1%)\n",
    "\n",
    "# Filter out reads with expected number of errors greater than 1.0: \\\n",
    "#     https://www.drive5.com/usearch/manual/exp_errs.html\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter within_p_phix_f.fastq -fastq_maxee 1.0 -fastaout within_p_filtered.fasta\n",
    "\n",
    "# 00:04 683Mb   100.0% Filtering, 96.6% passed\n",
    "#     571334  Reads (571.3k)                  \n",
    "#      19582  Discarded reads with expected errs > 1.00\n",
    "#     551752  Filtered reads (551.8k, 96.6%)\n",
    "\n",
    "# Re-orient the reads to forward, filter out reads with an undetermined orientation\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient within_p_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout \\\n",
    "    within_p_oriented.fasta -tabbedout within_p_oriented.txt\n",
    "\n",
    "# 00:04 1.0Gb   100.0% 548048 plus (99.3%), 2 minus (0.0%), 3702 undet. (0.7%)\n",
    "\n",
    "# Find unique sequences for denoising\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques within_p_oriented.fasta -fastaout within_p_uniques.fasta -sizeout \\\n",
    "    -relabel Uniq\n",
    "\n",
    "# 00:02 898Mb  548050 seqs, 20995 uniques, 15276 singletons (72.8%)\n",
    "# 00:02 898Mb  Min size 1, median 1, max 180839, avg 26.10\n",
    "# 00:03 898Mb   100.0% Writing within_p_uniques.fasta\n",
    "\n",
    "# De-noise reads into zero-radius otus (i.e. ESVs or ASVs)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 within_p_uniques.fasta -zotus within_p_zotus.fasta\n",
    "\n",
    "# 00:00 49Mb    100.0% Reading within_p_uniques.fasta\n",
    "# 00:00 30Mb    100.0% 620 amplicons, 38994 bad (size >= 8)\n",
    "# 00:05 41Mb    100.0% 615 good, 5 chimeras                \n",
    "# 00:05 41Mb    100.0% Writing zotus\n",
    "\n",
    "# Make otu table with original sequencing reads mapped to zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab within_p_phix_f.fastq -otus within_p_zotus.fasta -otutabout \\\n",
    "    within_p_zotutab2.txt -mapout within_p_zmap2.txt -sample_delim .\n",
    "\n",
    "# 03:32 726Mb   100.0% Searching, 99.1% matchedchedg, 98.6% matched\n",
    "# 566207 / 571334 mapped to OTUs (99.1%)\n",
    "# 03:32 726Mb  Writing within_p_zotutab.txt\n",
    "# 03:32 726Mb  Writing within_p_zotutab.txt ...done.\n",
    "\n",
    "# Assign taxonomy to the zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax within_p_zotus.fasta -db ../../../RDP/rdp_its_v2.udb -tabbedout \\\n",
    "    within_p_zotus.sintax -strand both\n",
    "\n",
    "# Output files for analyses in R:\n",
    "\n",
    "# 'within_p_zotutab2.txt'\n",
    "# 'within_p_map_file.txt'\n",
    "# 'within_p_zotus3.sintax'\n",
    "#     Note: For the sintax file, it was modified in a text editor from the following format \\\n",
    "#     (within_p_zotus.sintax)\n",
    "\n",
    "#     Zotu5\td:Fungi(1.0000),p:Basidiomycota(0.8600),c:Agaricomycetes(0.7700),o:Sebacinales(0.4700),f:\\\n",
    "#                         Sebacinales_Group_A(0.4700),g:Sebacina(0.4700),s:Sebacina_incrustans(0.4700)\t+\td:\\\n",
    "#                                     Fungi,p:Basidiomycota\n",
    "\n",
    "#     To the format below (within_p_zotus3.sintax)\n",
    "\n",
    "#     Zotu5\tFungi\tBasidiomycota\tAgaricomycetes\tSebacinales\tSebacinales_Group_A\tSebacina\t\\\n",
    "#     Sebacina_incrustans\n",
    " \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
