{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Informatics code for Younginger, Stewart, Balkan and Ballhorn\n",
    "## \"Stable coexistence or competitive exclusion? \n",
    "## Fern endophytes shift our understanding of microbial temporal turnover\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Software needed for analysis:\n",
    "FastQC v0.11.5\n",
    "usearch v10.0.247M_i86linux64\n",
    "jun_jul_renamer.sh\n",
    "control_renamer.sh\n",
    "R1_renamer.sh\n",
    "\n",
    "\n",
    "## Sample processing\n",
    "\n",
    "### Concatenate the R1 reads since they were demultiplexed:\n",
    "\n",
    "cat *_R1_* > ../FastQC/jun_jul_R1.fastq.gz\n",
    "\n",
    "##### Total raw R1 reads for each month #####\n",
    "\n",
    "- Jun_Jul: 2605987\n",
    "- Aug_Sep: 2440660\n",
    "- Oct_Nov: 3310806\n",
    "- Dec_Jan: 2465935\n",
    "- preprocessed Apr_May: 15566066\n",
    "- demultiplexed Apr_May: 7955646\n",
    "\n",
    "## Demultiplex April May reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_demux Undetermined_S0_L001_R1_001.fastq -reverse Undetermined_S0_L001_R2_001.fastq -index Undetermined_S0_L001_I1_001.fastq -barcodes barcodes_6.fa -fastqout apr_may_R1.fastq -output2 apr_may_R2.fastq\n",
    "\n",
    "\t03:30 110Mb   100.0% Demuxed 7955646 / 15566066 (51.1%)\n",
    "\n",
    "## Consider using only R1 reads and also merging, running the analyses side-by-side to see how the results differ...\n",
    "\t## Will merge first though\n",
    "\n",
    "## Also, need to look at how the files are labeled and rename the control files to match the appropriate formatting\n",
    "\t## Jun_Jul will need to be renamed to match the other months, too!\n",
    "\n",
    "# Changed the Jun_Jul samples with the following script:\n",
    "\n",
    "jun_jul_renamer.sh\n",
    "\n",
    "# Will now change up the names of the controls in the remaining samples\n",
    "\n",
    "control_renamer.sh\n",
    "\n",
    "## Merging reads\n",
    "\n",
    "## Apr_May\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs apr_may_R1.fastq -fastqout ../Merged_reads/apr_may.fastq -fastq_maxdiffs 300\n",
    "\n",
    "\n",
    "Merging\n",
    "  Fwd apr_may_R1.fastq\n",
    "  Rev apr_may_R2.fastq\n",
    "  Keep read labels\n",
    "\n",
    "00:01 71Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "01:14 751Mb   100.0% 91.4% merged\n",
    "\n",
    "Totals:\n",
    "   7955646  Pairs (8.0M)\n",
    "   7274761  Merged (7.3M, 91.44%)\n",
    "   1423348  Alignments with zero diffs (17.89%)\n",
    "    641535  Too many diffs (> 300) (8.06%)\n",
    "     39350  No alignment found (0.49%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   7469521  Staggered pairs (93.89%) merged & trimmed\n",
    "    205.78  Mean alignment length\n",
    "    209.40  Mean merged length\n",
    "      0.33  Mean fwd expected errors\n",
    "      1.90  Mean rev expected errors\n",
    "      0.14  Mean merged expected errors\n",
    "      \n",
    "## Jun_Jul\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/jun_jul.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:18 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:18 765Mb   100.0% 91.9% merged\n",
    "\n",
    "Totals:\n",
    "   2605987  Pairs (2.6M)\n",
    "   2395636  Merged (2.4M, 91.93%)\n",
    "    193274  Alignments with zero diffs (7.42%)\n",
    "    175297  Too many diffs (> 300) (6.73%)\n",
    "     35054  No alignment found (1.35%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2344067  Staggered pairs (89.95%) merged & trimmed\n",
    "    199.25  Mean alignment length\n",
    "    216.07  Mean merged length\n",
    "      0.64  Mean fwd expected errors\n",
    "      1.70  Mean rev expected errors\n",
    "      0.19  Mean merged expected errors\n",
    "\n",
    "#### Wondering if I need to deal with differences in header lines between Apr_May and the rest of the files:\n",
    "\n",
    "@M02149:71:000000000-AAMFM:1:1101:13889:1730;sample=BY-12-1-5;\n",
    "\n",
    "@BY-10-1-6.2\n",
    "\n",
    "## Will try to address it with the following argument: -relabel @\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs apr_may_R1.fastq -fastqout ../Merged_reads/apr_may_V2.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\t# Nope, that didn't do it. Hopefully it will get taken care of later in processing  \n",
    "    \n",
    "## Aug_Sep\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/aug_sep.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:15 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:15 765Mb   100.0% 54.8% merged\n",
    "\n",
    "Totals:\n",
    "   2440660  Pairs (2.4M)\n",
    "   1337858  Merged (1.3M, 54.82%)\n",
    "      9220  Alignments with zero diffs (0.38%)\n",
    "   1083301  Too many diffs (> 300) (44.39%)\n",
    "         4  Fwd tails Q <= 2 trimmed (0.00%)\n",
    "        41  Rev tails Q <= 2 trimmed (0.00%)\n",
    "        87  Fwd too short (< 64) after tail trimming (0.00%)\n",
    "       309  Rev too short (< 64) after tail trimming (0.01%)\n",
    "     19105  No alignment found (0.78%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2306121  Staggered pairs (94.49%) merged & trimmed\n",
    "    201.15  Mean alignment length\n",
    "    210.86  Mean merged length\n",
    "      1.16  Mean fwd expected errors\n",
    "      3.87  Mean rev expected errors\n",
    "      0.62  Mean merged expected errors\n",
    "\n",
    "# Too many diffs resulted in 44.39% of the reads getting discarded! Did this happen before? \n",
    "\t# May need to just continue with the R1s...\n",
    "\t# But what if I run it with 400 max diffs? \n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/aug_sep_V2.fastq -relabel @ -fastq_maxdiffs 400\n",
    "\t# Results in the exact same output. Really weird and an indication that I should be using the R1 reads only\n",
    "\n",
    "## Oct_Nov\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout oct_nov.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\t# Need to move this into the Merged_reads directory since I didn't specify\n",
    "\n",
    "00:21 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:21 765Mb   100.0% 90.4% merged\n",
    "\n",
    "Totals:\n",
    "   3310806  Pairs (3.3M)\n",
    "   2992916  Merged (3.0M, 90.40%)\n",
    "     34999  Alignments with zero diffs (1.06%)\n",
    "    306191  Too many diffs (> 300) (9.25%)\n",
    "     11699  No alignment found (0.35%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2946717  Staggered pairs (89.00%) merged & trimmed\n",
    "    197.32  Mean alignment length\n",
    "    219.33  Mean merged length\n",
    "      0.64  Mean fwd expected errors\n",
    "      2.59  Mean rev expected errors\n",
    "      0.58  Mean merged expected errors\n",
    "\n",
    "## Dec_Jan\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/dec_jan.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "00:16 765Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:16 765Mb   100.0% 94.2% merged\n",
    "\n",
    "Totals:\n",
    "   2465935  Pairs (2.5M)\n",
    "   2323617  Merged (2.3M, 94.23%)\n",
    "     63861  Alignments with zero diffs (2.59%)\n",
    "    133325  Too many diffs (> 300) (5.41%)\n",
    "      8993  No alignment found (0.36%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "   2158636  Staggered pairs (87.54%) merged & trimmed\n",
    "    196.24  Mean alignment length\n",
    "    221.91  Mean merged length\n",
    "      0.80  Mean fwd expected errors\n",
    "      2.38  Mean rev expected errors\n",
    "      0.33  Mean merged expected errors\n",
    "\n",
    "## Put them all together\n",
    "\n",
    "cat *.fastq > temporal_merged.fastq\n",
    "\n",
    "## Now look into renaming the R1 files and concatenating them\n",
    "\n",
    "## Need to relabel all the header lines with the fastx_relabel command with usearch\n",
    "\t## This should be easy enough with a shell script, but need to deal with the Jun_Jul samples\n",
    "\t## and the controls. I should make a separate script for each month-group\n",
    "\t\n",
    "# Used the following script to rename the R1s only:\n",
    "\n",
    "R1_renamer.sh\n",
    "\n",
    "# Concerned about the differences in the header lines between Apr_May and the rest, but will have\n",
    "\t# to wait and see how usearch handles it\n",
    "\t\n",
    "cat * > ../../Forward_reads/jun_jul_R1.fastq\n",
    "\t# Did the same for the remaining months\n",
    "\n",
    "# Okay, the merged and R1 files are in the following directories, respectively:\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Merged_reads/temporal_merged.fastq\n",
    "\t### Read total: 16324788 ###\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Forward_reads/temporal_R1.fastq\n",
    "\t### Read total: 18779034 ###\n",
    "\n",
    "# Will work on Cutadapt next:\n",
    "\n",
    "## R1 reads only:\n",
    "\n",
    "cutadapt -a GATCTCTTGGNTCTNGCATCGATGAAGAACG -q 20 -e 0.2 temporal_R1.fastq -o temporal_R1_cutadapt.fastq\n",
    "\n",
    "This is cutadapt 1.10 with Python 2.7.7\n",
    "Command line parameters: -a GATCTCTTGGNTCTNGCATCGATGAAGAACG -q 20 -e 0.2 temporal_R1.fastq -o temporal_R1_cutadapt.fastq\n",
    "Trimming 1 adapter with at most 20.0% errors in single-end mode ...\n",
    "\n",
    "\n",
    "Finished in 474.03 s (25 us/read; 2.38 M reads/minute).\n",
    "\n",
    "=== Summary ===\n",
    "\n",
    "Total reads processed:              18,779,034\n",
    "Reads with adapters:                17,002,958 (90.5%)\n",
    "Reads written (passing filters):    18,779,034 (100.0%)\n",
    "\n",
    "Total basepairs processed: 4,709,405,469 bp\n",
    "Quality-trimmed:              56,305,856 bp (1.2%)\n",
    "Total written (filtered):  3,905,674,514 bp (82.9%)\n",
    "\n",
    "=== Adapter 1 ===\n",
    "\n",
    "Sequence: GATCTCTTGGNTCTNGCATCGATGAAGAACG; Type: regular 3'; Length: 31; Trimmed: 17002958 times.\n",
    "\n",
    "No. of allowed errors:\n",
    "0-4 bp: 0; 5-9 bp: 1; 10-14 bp: 2; 15-19 bp: 3; 20-24 bp: 4; 25-29 bp: 5; 30-31 bp: 6\n",
    "\n",
    "Bases preceding removed adapters:\n",
    "  A: 0.3%\n",
    "  C: 0.3%\n",
    "  G: 98.6%\n",
    "  T: 0.9%\n",
    "  none/other: 0.0%\n",
    "WARNING:\n",
    "    The adapter is preceded by \"G\" extremely often.\n",
    "    The provided adapter sequence may be incomplete.\n",
    "    To fix the problem, add \"G\" to the beginning of the adapter sequence.\n",
    "\n",
    "Overview of removed sequences\n",
    "length\tcount\texpect\tmax.err\terror counts\n",
    "3\t31375\t293422.4\t0\t31375\n",
    "4\t28238\t73355.6\t0\t5937 22301\n",
    "5\t22333\t18338.9\t1\t7998 14335\n",
    "6\t6189\t4584.7\t1\t2531 3658\n",
    "7\t22495\t1146.2\t1\t18857 3638\n",
    "8\t47008\t286.5\t1\t40986 5554 468\n",
    "9\t8274\t71.6\t1\t5165 1596 1513\n",
    "10\t7200\t17.9\t2\t3041 1185 2974\n",
    "11\t7357\t4.5\t2\t4212 1968 1177\n",
    "12\t16153\t1.1\t2\t11227 2929 1952 45\n",
    "13\t7455\t0.3\t2\t4252 1870 993 340\n",
    "14\t27889\t0.1\t2\t18397 6911 2338 243\n",
    "15\t17096\t0.0\t3\t10683 4247 1596 570\n",
    "16\t26221\t0.0\t3\t18821 4962 1653 784 1\n",
    "17\t27322\t0.0\t3\t19179 5166 1657 1061 259\n",
    "18\t12899\t0.0\t3\t8406 2558 1119 472 344\n",
    "19\t23484\t0.0\t3\t16584 3758 1673 665 804\n",
    "20\t45336\t0.0\t4\t27816 11078 4132 1527 781 2\n",
    "21\t52374\t0.0\t4\t27228 14602 3381 1243 5917 3\n",
    "22\t21090\t0.0\t4\t6973 6599 1875 904 4643 96\n",
    "23\t50084\t0.0\t4\t31639 11956 3998 1599 870 22\n",
    "24\t73617\t0.0\t4\t48967 15172 6296 2016 1082 78 6\n",
    "25\t50452\t0.0\t5\t35043 9472 3187 1498 718 450 84\n",
    "26\t263347\t0.0\t5\t224053 29662 5620 2044 1133 739 96\n",
    "27\t194213\t0.0\t5\t158479 23443 5682 2284 2518 1373 434\n",
    "28\t32278\t0.0\t5\t14878 9939 4171 1836 978 450 26\n",
    "29\t98665\t0.0\t5\t58998 19428 8475 4134 5962 1630 38\n",
    "30\t53098\t0.0\t6\t21977 12658 6444 3048 7422 1165 384\n",
    "31\t45000\t0.0\t6\t21867 11834 6037 2779 1468 663 352\n",
    "32\t70381\t0.0\t6\t37861 17673 8230 3811 1644 780 382\n",
    "33\t69273\t0.0\t6\t38613 16868 7638 3392 1603 740 419\n",
    "34\t68984\t0.0\t6\t31130 15773 7605 5557 3923 2949 2047\n",
    "35\t87784\t0.0\t6\t47293 20269 8094 5650 3210 2079 1189\n",
    "36\t260133\t0.0\t6\t156882 57877 20757 12729 6437 3657 1794\n",
    "37\t913918\t0.0\t6\t703775 123618 35418 23912 10373 5127 11695\n",
    "38\t407887\t0.0\t6\t315044 53881 15420 10661 4746 2512 5623\n",
    "39\t255709\t0.0\t6\t197932 33309 9528 6523 2912 1697 3808\n",
    "40\t227071\t0.0\t6\t176411 28766 8478 5534 2515 1515 3852\n",
    "41\t226038\t0.0\t6\t174433 28484 8465 5264 2401 1519 5472\n",
    "42\t232753\t0.0\t6\t182365 28607 8154 5525 2642 1690 3770\n",
    "43\t371030\t0.0\t6\t296476 41970 11727 8155 3838 2487 6377\n",
    "44\t572833\t0.0\t6\t457233 61932 18349 11896 5632 3616 14175\n",
    "45\t867901\t0.0\t6\t702918 93397 25678 18176 8691 5868 13173\n",
    "46\t3276664\t0.0\t6\t2772634 293239 72195 50987 23560 16056 47993\n",
    "47\t6077651\t0.0\t6\t5264534 454810 110150 77091 35670 25385 110011\n",
    "48\t35061\t0.0\t6\t20401 5610 1982 1621 1468 1932 2047\n",
    "49\t43796\t0.0\t6\t31231 7787 2047 1155 665 450 461\n",
    "50\t89724\t0.0\t6\t72792 10909 2092 2001 808 535 587\n",
    "51\t831068\t0.0\t6\t748778 62519 7447 8464 1487 865 1508\n",
    "52\t50675\t0.0\t6\t38116 7338 1825 1258 680 711 747\n",
    "53\t36339\t0.0\t6\t30508 3970 817 283 190 136 435\n",
    "54\t32163\t0.0\t6\t25122 4809 1033 379 208 167 445\n",
    "55\t13365\t0.0\t6\t10637 1654 358 265 140 127 184\n",
    "56\t101183\t0.0\t6\t87445 10080 1291 1342 531 218 276\n",
    "57\t15735\t0.0\t6\t11335 3137 567 265 126 191 114\n",
    "58\t99392\t0.0\t6\t87135 8479 1450 915 495 416 502\n",
    "59\t7373\t0.0\t6\t5437 1056 301 210 105 113 151\n",
    "60\t15062\t0.0\t6\t12341 1767 358 226 124 128 118\n",
    "61\t29378\t0.0\t6\t24613 3249 604 342 169 167 234\n",
    "62\t8251\t0.0\t6\t6638 809 237 169 112 116 170\n",
    "63\t9367\t0.0\t6\t7864 819 199 186 107 75 117\n",
    "64\t2628\t0.0\t6\t1326 687 208 142 99 85 81\n",
    "65\t23272\t0.0\t6\t5081 14903 1863 562 471 219 173\n",
    "66\t44629\t0.0\t6\t31068 7110 1902 1920 1315 755 559\n",
    "67\t7952\t0.0\t6\t2744 2080 811 849 649 489 330\n",
    "68\t11784\t0.0\t6\t8341 2245 542 272 161 82 141\n",
    "69\t4656\t0.0\t6\t3109 948 232 142 77 64 84\n",
    "70\t1081\t0.0\t6\t451 285 118 54 50 57 66\n",
    "71\t45798\t0.0\t6\t36741 6419 887 534 470 407 340\n",
    "72\t44880\t0.0\t6\t33993 6193 1815 915 535 425 1004\n",
    "73\t3502\t0.0\t6\t2669 463 162 98 48 33 29\n",
    "74\t1742\t0.0\t6\t1278 234 72 78 30 26 24\n",
    "75\t2289\t0.0\t6\t1709 327 77 76 33 29 38\n",
    "76\t8822\t0.0\t6\t6223 1890 284 175 80 55 115\n",
    "77\t11994\t0.0\t6\t8002 2661 604 262 196 169 100\n",
    "78\t2080\t0.0\t6\t1579 274 118 39 21 23 26\n",
    "79\t18698\t0.0\t6\t13978 3910 443 130 133 61 43\n",
    "80\t2896\t0.0\t6\t2332 376 91 47 24 15 11\n",
    "81\t1384\t0.0\t6\t801 345 156 41 16 11 14\n",
    "82\t3716\t0.0\t6\t2039 1161 355 76 39 29 17\n",
    "83\t10655\t0.0\t6\t6310 3183 874 149 54 27 58\n",
    "84\t6795\t0.0\t6\t4413 1579 573 108 48 31 43\n",
    "85\t2276\t0.0\t6\t1419 579 157 43 48 16 14\n",
    "86\t1187\t0.0\t6\t749 279 95 32 16 5 11\n",
    "87\t803\t0.0\t6\t455 165 77 28 29 20 29\n",
    "88\t1420\t0.0\t6\t761 265 121 97 68 68 40\n",
    "89\t275\t0.0\t6\t175 56 21 7 6 4 6\n",
    "90\t422\t0.0\t6\t299 79 16 10 7 8 3\n",
    "91\t286\t0.0\t6\t192 49 14 13 7 7 4\n",
    "92\t634\t0.0\t6\t492 73 23 13 15 9 9\n",
    "93\t109\t0.0\t6\t62 34 8 2 1 0 2\n",
    "94\t142\t0.0\t6\t84 34 13 2 4 2 3\n",
    "95\t239\t0.0\t6\t130 58 18 9 8 8 8\n",
    "96\t488\t0.0\t6\t308 91 37 19 9 12 12\n",
    "97\t324\t0.0\t6\t152 119 16 10 5 11 11\n",
    "98\t797\t0.0\t6\t475 153 44 50 33 23 19\n",
    "99\t403\t0.0\t6\t32 276 52 15 11 11 6\n",
    "100\t77\t0.0\t6\t29 34 7 5 0 1 1\n",
    "101\t119\t0.0\t6\t45 49 13 3 4 3 2\n",
    "102\t207\t0.0\t6\t126 51 12 5 6 4 3\n",
    "103\t545\t0.0\t6\t388 100 23 8 11 2 13\n",
    "104\t78\t0.0\t6\t36 25 8 4 1 2 2\n",
    "105\t130\t0.0\t6\t52 51 11 6 4 2 4\n",
    "106\t197\t0.0\t6\t68 84 29 4 2 4 6\n",
    "107\t190\t0.0\t6\t87 54 36 4 2 2 5\n",
    "108\t205\t0.0\t6\t95 66 31 5 4 1 3\n",
    "109\t365\t0.0\t6\t85 134 108 16 12 5 5\n",
    "110\t957\t0.0\t6\t169 397 303 48 11 13 16\n",
    "111\t2302\t0.0\t6\t523 1372 248 71 39 26 23\n",
    "112\t1823\t0.0\t6\t1368 240 77 53 18 14 53\n",
    "113\t22\t0.0\t6\t17 2 0 1 1 1\n",
    "114\t15\t0.0\t6\t11 4\n",
    "115\t22\t0.0\t6\t16 4 0 0 0 2\n",
    "116\t31\t0.0\t6\t26 4 0 0 0 0 1\n",
    "117\t40\t0.0\t6\t29 4 1 3 0 0 3\n",
    "118\t127\t0.0\t6\t104 20 1 1 0 1\n",
    "119\t361\t0.0\t6\t296 50 6 5 1 2 1\n",
    "120\t24\t0.0\t6\t21 2 1\n",
    "121\t8\t0.0\t6\t4 2 0 0 1 0 1\n",
    "122\t5\t0.0\t6\t0 2 3\n",
    "123\t6\t0.0\t6\t5 0 0 0 0 1\n",
    "124\t1\t0.0\t6\t1\n",
    "125\t6\t0.0\t6\t3 0 1 0 1 1\n",
    "126\t9\t0.0\t6\t7 1 0 1\n",
    "127\t14\t0.0\t6\t7 5 1 1\n",
    "128\t15\t0.0\t6\t7 2 4 0 0 0 2\n",
    "129\t6\t0.0\t6\t5 0 1\n",
    "130\t14\t0.0\t6\t13 0 1\n",
    "131\t12\t0.0\t6\t8 1 1 0 0 2\n",
    "132\t26\t0.0\t6\t19 2 4 0 0 0 1\n",
    "133\t32\t0.0\t6\t23 4 3 0 0 0 2\n",
    "134\t33\t0.0\t6\t25 4 4\n",
    "135\t115\t0.0\t6\t89 15 4 4 1 1 1\n",
    "136\t393\t0.0\t6\t337 27 16 9 1 1 2\n",
    "137\t17\t0.0\t6\t6 0 11\n",
    "138\t65\t0.0\t6\t13 1 46 4 0 1\n",
    "139\t149\t0.0\t6\t13 0 124 10 1 1\n",
    "140\t31\t0.0\t6\t27 1 0 1 1 1\n",
    "141\t90\t0.0\t6\t83 5 1 0 0 1\n",
    "142\t262\t0.0\t6\t223 25 6 2 1 2 3\n",
    "143\t23\t0.0\t6\t18 4 0 0 1\n",
    "144\t45\t0.0\t6\t33 10 0 0 1 0 1\n",
    "145\t9\t0.0\t6\t7 1 1\n",
    "146\t9\t0.0\t6\t8 1\n",
    "147\t37\t0.0\t6\t34 2 1\n",
    "148\t99\t0.0\t6\t91 3 3 0 0 0 2\n",
    "149\t3\t0.0\t6\t3\n",
    "151\t1\t0.0\t6\t0 1\n",
    "152\t2\t0.0\t6\t0 2\n",
    "153\t1\t0.0\t6\t0 1\n",
    "156\t1\t0.0\t6\t0 1\n",
    "157\t1\t0.0\t6\t0 1\n",
    "158\t2\t0.0\t6\t0 2\n",
    "159\t5\t0.0\t6\t0 5\n",
    "160\t15\t0.0\t6\t0 15\n",
    "164\t1\t0.0\t6\t0 1\n",
    "176\t2\t0.0\t6\t2\n",
    "192\t1\t0.0\t6\t0 1\n",
    "196\t4\t0.0\t6\t0 4\n",
    "197\t8\t0.0\t6\t0 8\n",
    "209\t1\t0.0\t6\t0 0 0 0 0 0 1\n",
    "\n",
    "WARNING:\n",
    "    One or more of your adapter sequences may be incomplete.\n",
    "    Please see the detailed output above.\n",
    "\n",
    "### For paired reads which always overlap (per usearch docs):\n",
    "If the read length is long enough that the longest ITS sequence will given an overlap of at least, say, \n",
    "32 bases, then you don't need any additional trimming:  fastq_mergepairs does everything you need. Short amplicons will \n",
    "create \"staggered\" pairs which are correctly truncated during the merging.\"\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Merged_reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter temporal_merged.fastq -fastq_maxee 1.0 -fastaout temporal_merged_filtered.fasta\n",
    "\n",
    "00:00 38Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "01:37 683Mb   100.0% Filtering, 93.8% passed\n",
    "  16324788  Reads (16.3M)                   \n",
    "   1007674  Discarded reads with expected errs > 1.00\n",
    "  15317114  Filtered reads (15.3M, 93.8%)\n",
    "  \n",
    "/workspace/scratch/obrett/Temporal/Forward_reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter temporal_R1_cutadapt.fastq -fastq_maxee 1.0 -fastaout temporal_R1_filtered.fasta\n",
    "\n",
    "00:01 38Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "02:19 683Mb   100.0% Filtering, 83.9% passed\n",
    "  18779034  Reads (18.8M)                   \n",
    "   3031410  Discarded reads with expected errs > 1.00\n",
    "  15747604  Filtered reads (15.7M, 83.9%)\n",
    "\n",
    "# Okay, 15.3M reads in the merged dataset, 15.7M reads in the R1 dataset.\n",
    "\t# Looks like those same reads in the Aug_Sep group were kicked out\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques Merged_reads/temporal_merged_filtered.fasta -fastaout OTUs/merged_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:44 4.7Gb   100.0% Reading Merged_reads/temporal_merged_filtered.fasta\n",
    "00:44 4.7Gb  CPU has 28 cores, defaulting to 10 threads                 \n",
    "00:54 8.2Gb   100.0% DF\n",
    "01:03 8.3Gb  15317114 seqs, 802957 uniques, 660362 singletons (82.2%)\n",
    "01:03 8.3Gb  Min size 1, median 1, max 9957620, avg 19.08\n",
    "01:10 5.8Gb   100.0% Writing OTUs/merged_uniques.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques Forward_reads/temporal_R1_filtered.fasta -fastaout OTUs/R1_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:46 4.8Gb   100.0% Reading Forward_reads/temporal_R1_filtered.fasta\n",
    "00:46 4.7Gb  CPU has 28 cores, defaulting to 10 threads              \n",
    "00:56 8.3Gb   100.0% DF\n",
    "01:05 8.4Gb  15747604 seqs, 958884 uniques, 757129 singletons (79.0%)\n",
    "01:05 8.4Gb  Min size 1, median 1, max 9174463, avg 16.42\n",
    "01:12 5.9Gb   100.0% Writing OTUs/R1_uniques.fasta\n",
    "\n",
    "## 2020-02-12\n",
    "\n",
    "# Will cluster OTUs and denoise into ZOTUs for both the merged and R1 reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -cluster_otus merged_uniques.fasta -otus merged_otus.fasta -relabel Otu\n",
    "\n",
    "00:17 61Mb    100.0% 1301 OTUs, 320 chimeras\n",
    "\n",
    "usearch10.0.274M_i86linux64 -cluster_otus R1_uniques.fasta -otus R1_otus.fasta -relabel Otu\n",
    "\n",
    "00:19 60Mb    100.0% 1474 OTUs, 670 chimeras\n",
    "\t\n",
    "# Already 170 more OTUs and 350 more chimeras in the R1 dataset\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 merged_uniques.fasta -zotus merged_zotus.fasta\n",
    "\n",
    "00:03 333Mb   100.0% Reading merged_uniques.fasta\n",
    "00:04 327Mb   100.0% 992 amplicons, 2636496 bad (size >= 8) \n",
    "00:20 338Mb   100.0% 985 good, 7 chimeras                  \n",
    "00:20 338Mb   100.0% Writing zotus\n",
    "\n",
    "# Wow, already > 300 fewer ZOTUs than OTUs\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 R1_uniques.fasta -zotus R1_zotus.fasta\n",
    "\n",
    "00:02 316Mb   100.0% Reading R1_uniques.fasta\n",
    "00:04 316Mb   100.0% 1187 amplicons, 3803606 bad (size >= 8) \n",
    "00:21 327Mb   100.0% 1180 good, 7 chimeras                  \n",
    "00:21 327Mb   100.0% Writing zotus\n",
    "\n",
    "# For constructing these tables, I'll want to use the following files: \n",
    "\ttemporal_merged.fastq\n",
    "\ttemporal_R1_cutadapt.fastq\n",
    "\t\n",
    "usearch10.0.274M_i86linux64 -otutab ../Merged_reads/temporal_merged.fastq -otus merged_otus.fasta -otutabout merged_otutab.txt -mapout merged_map.txt\n",
    "\n",
    "00:00 43Mb    100.0% Reading merged_otus.fasta\n",
    "00:00 9.3Mb   100.0% Masking (fastnucleo)\n",
    "00:00 10Mb    100.0% Word stats\n",
    "00:00 10Mb    100.0% Alloc rows\n",
    "00:00 12Mb    100.0% Build index\n",
    "00:00 45Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "52:57 737Mb   100.0% Searching temporal_merged.fastq, 99.8% matched\n",
    "16291012 / 16324788 mapped to OTUs (99.8%)\n",
    "52:57 737Mb  Writing merged_otutab.txt\n",
    "52:57 737Mb  Writing merged_otutab.txt ...done.\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab ../Forward_reads/temporal_R1_cutadapt.fastq -otus R1_otus.fasta -otutabout R1_otutab.txt -mapout R1_map.txt\n",
    "\n",
    "00:00 43Mb    100.0% Reading R1_otus.fasta\n",
    "00:00 9.3Mb   100.0% Masking (fastnucleo)\n",
    "00:00 11Mb    100.0% Word stats\n",
    "00:00 11Mb    100.0% Alloc rows\n",
    "00:00 12Mb    100.0% Build index\n",
    "00:00 45Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "53:02 736Mb   100.0% Searching temporal_R1_cutadapt.fastq, 96.0% matched\n",
    "18021157 / 18779034 mapped to OTUs (96.0%)\n",
    "53:02 736Mb  Writing R1_otutab.txt\n",
    "53:02 736Mb  Writing R1_otutab.txt ...done.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab ../Merged_reads/temporal_merged.fastq -otus merged_zotus.fasta -otutabout merged_zotutab.txt -mapout merged_zmap.txt\n",
    "\n",
    "00:00 43Mb    100.0% Reading merged_zotus.fasta\n",
    "00:00 9.3Mb   100.0% Masking (fastnucleo)\n",
    "00:00 10Mb    100.0% Word stats\n",
    "00:00 10Mb    100.0% Alloc rows\n",
    "00:00 11Mb    100.0% Build index\n",
    "00:00 44Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "53:22 726Mb   100.0% Searching temporal_merged.fastq, 94.8% matched\n",
    "15471210 / 16324788 mapped to OTUs (94.8%)\n",
    "53:22 726Mb  Writing merged_zotutab.txt\n",
    "53:22 726Mb  Writing merged_zotutab.txt ...done.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab ../Forward_reads/temporal_R1_cutadapt.fastq -otus R1_zotus.fasta -otutabout R1_zotutab.txt -mapout R1_zmap.txt\n",
    "\n",
    "00:00 43Mb    100.0% Reading R1_zotus.fasta\n",
    "00:00 9.3Mb   100.0% Masking (fastnucleo)\n",
    "00:00 10Mb    100.0% Word stats\n",
    "00:00 10Mb    100.0% Alloc rows\n",
    "00:00 11Mb    100.0% Build index\n",
    "00:00 45Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "56:46 736Mb   100.0% Searching temporal_R1_cutadapt.fastq, 95.7% matched\n",
    "17979390 / 18779034 mapped to OTUs (95.7%)                              \n",
    "56:46 736Mb  Writing R1_zotutab.txt\n",
    "56:46 736Mb  Writing R1_zotutab.txt ...done.\n",
    "\n",
    "### Okay, think this is done for now. Will want to go back and run some of the QC checks, but will look at the files now:\n",
    "\n",
    "# Also, should make a tarball of the Temporal/ directory now, too\n",
    "\n",
    "## Overview of final outputs:\n",
    "\n",
    "- merged otus: 16291012 / 16324788 mapped to OTUs (99.8%) - 1301 OTUs, 320 chimeras\n",
    "- R1 otus: 18021157 / 18779034 mapped to OTUs (96.0%) - 1474 OTUs, 670 chimeras\n",
    "- merged zotus: 15471210 / 16324788 mapped to OTUs (94.8%) - 985 good, 7 chimeras\n",
    "- R1 zotus: 17979390 / 18779034 mapped to OTUs (95.7%) - 1180 good, 7 chimeras\n",
    "\n",
    "## Would like to explore the crosstalk option now:\n",
    "\t## Instead, think I'm going to go through the QC pipeline Robert Edgar recommends:\n",
    "\t\thttps://www.drive5.com/usearch/manual/otu_qc.html\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/OTUs/QC\n",
    "\n",
    "## Suppose I should download and get the most recent UNITE database, too. \n",
    "\n",
    "/workspace/scratch/obrett/UNITE/uchime_reference_dataset_28.06.2017/ITS1_ITS2_datasets\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax uchime_reference_dataset_ITS1_28.06.2017.fasta -output ~/UNITE2/ITS1_28_06_2017.udb\n",
    "\t\n",
    "00:00 55Mb    100.0% Reading uchime_reference_dataset_ITS1_28.06.2017.fasta\n",
    "00:00 22Mb    100.0% Converting to upper case                              \n",
    "00:00 23Mb    100.0% Word stats              \n",
    "00:00 23Mb    100.0% Alloc rows\n",
    "00:01 48Mb    100.0% Build index\n",
    "00:01 50Mb    100.0% Initialize taxonomy data\n",
    "00:01 50Mb    100.0% Building name table     \n",
    "00:01 50Mb   1 names, tax levels min 0, avg 0.0, max 0\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax uchime_reference_dataset_ITS1_28.06.2017.fasta -output /home/obrett/UNITE2/ITS1_28_06_2017.udb\n",
    "\n",
    "---Fatal error---\n",
    "../utaxdata.cpp(949) assert failed: Size > 0 && Ids != 0\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax uchime_reference_dataset_28.06.2017.fasta -output ~/UNITE2/ITS_2017_06_28.udb\n",
    "\n",
    "00:00 65Mb    100.0% Reading uchime_reference_dataset_28.06.2017.fasta\n",
    "00:00 32Mb    100.0% Converting to upper case                         \n",
    "00:01 33Mb    100.0% Word stats              \n",
    "00:01 33Mb    100.0% Alloc rows\n",
    "00:02 98Mb    100.0% Build index\n",
    "00:02 100Mb   100.0% Initialize taxonomy data\n",
    "00:02 100Mb   100.0% Building name table     \n",
    "00:02 100Mb  1 names, tax levels min 0, avg 0.0, max 0\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax uchime_reference_dataset_28.06.2017.fasta -output /home/obrett/UNITE2/ITS_2017_06_28.udb\n",
    "\n",
    "---Fatal error---\n",
    "../utaxdata.cpp(949) assert failed: Size > 0 && Ids != 0\n",
    "\n",
    "# Need to figure out the error and get a new DB formatted and in the correct directory\n",
    "\n",
    "# 2020-02-13\n",
    "\n",
    "# Downloaded a different copy of an ITS dataset. Hopefully this is the right one.\n",
    "\n",
    "/workspace/scratch/obrett/UNITE\n",
    "\n",
    "wget https://files.plutof.ut.ee/public/orig/E8/83/E883EB19E3EA7B64C1F652521301239831FAFE0BFF015C9E2B4786DC0976C0FC.gz\n",
    "\n",
    "gunzip E883EB19E3EA7B64C1F652521301239831FAFE0BFF015C9E2B4786DC0976C0FC.gz\n",
    "\n",
    "mv E883EB19E3EA7B64C1F652521301239831FAFE0BFF015C9E2B4786DC0976C0FC utax_reference_dataset_04.02.2020.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax utax_reference_dataset_04.02.2020.fasta -output ~/UNITE2/unite_2020_02_04.udb\n",
    "\n",
    "---Fatal error---\n",
    "Missing x: in name >MF167586|SH1151072.08FU;tax=d:Fungi,p:Ascomycota,c:Sordariomycetes,o:Xylariales,f:,g:Delonicicola,s:Delonicicola_siamense_SH1151072.08FU;\n",
    "\n",
    "## Not sure if this is the only header in the entire file that is missing info, but the \"f:\" designation is missing. \n",
    "\t## This is the perfect time to use a python script to read the file, find the line that is messed up and replace it with the appropriate line\n",
    "\n",
    "udb_cleanup.py\n",
    "\n",
    "# Had to run the above twice to get the taxonomy file cleaned up, but did get this warning:\n",
    "\n",
    "WARNING: 7 taxonomy nodes have >1 parent\n",
    "\n",
    "# Will have to wait and see if this is a problem downstream. Still, the taxonomy file is here:\n",
    "\n",
    "/home/obrett/UNITE2/unite_2020_02_04.udb\n",
    "\n",
    "## Back to the QC stuff...\n",
    "\n",
    "\t# Making a subsample of the most and least abundant otus\n",
    "\n",
    "head -21 merged_otus.fasta > QC/alignment_subsample.fasta\n",
    "\n",
    "tail -19 merged_otus.fasta > QC/alignment_subsample_V2.fasta\n",
    "\n",
    "cat alignment_subsample.fasta alignment_subsample_V2.fasta > alignment_subsample_V3.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -usearch_global alignment_subsample_V3.fasta -db ~/UNITE2/unite_2020_02_04.udb -id 0.9 -strand both -alnout out.aln -uc otu.uc\n",
    "\n",
    "## I think the alignments look really good. Any differences are indels in the middle of the reads, not towards the ends. \n",
    "\t# Safe to proceed\n",
    "\t\n",
    "cut -f1 ../merged_otutab.txt | grep -v \"^#\" > table_labels.txt\n",
    "\n",
    "grep \"^>\" ../merged_otus.fasta | sed \"-es/>//\" > seq_labels.txt\n",
    "\n",
    "sort seq_labels.txt table_labels.txt table_labels.txt | uniq -u > missing_labels.txt\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_getseqs ../merged_otus.fasta -labels missing_labels.txt -fastaout missing.fasta\n",
    "\n",
    "00:00 3.6Mb  Reading missing_labels.txt...done.\n",
    "00:00 37Mb    100.0% Searching, 0 found\n",
    "\n",
    "## Safe to proceed\n",
    "\n",
    "## Looking at the orientation of the reads: \n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient ../merged_otus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout orient.txt\n",
    "\n",
    "cut -f2 orient.txt | sort | uniq -c\n",
    "     16 -\n",
    "    203 ?\n",
    "   1082 +\n",
    "\n",
    "# Wonder why 16 and 203 are minus and undetermined...\n",
    "\n",
    "# Okay, need to orient them before the command below:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques Merged_reads/temporal_merged_filtered.fasta -fastaout OTUs/merged_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "# Will do the following, then feed it back into the pipeline:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient ../../Merged_reads/temporal_merged_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout ../../Merged_reads/merged_oriented.fasta -tabbedout orient_V2.txt\n",
    "\n",
    "01:53 1.0Gb   100.0% 14687508 plus (95.9%), 12353 minus (0.1%), 617253 undet. (4.0%)\n",
    "\t# I wonder what this does with the undetermined reads???\n",
    "\n",
    "### Pipeline after this step:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques Merged_reads/temporal_merged_filtered.fasta -fastaout OTUs/merged_uniques.fasta -sizeout -relabel Uniq\n",
    "usearch10.0.274M_i86linux64 -cluster_otus merged_uniques.fasta -otus merged_otus.fasta -relabel Otu\n",
    "usearch10.0.274M_i86linux64 -otutab ../Merged_reads/temporal_merged.fastq -otus merged_otus.fasta -otutabout merged_otutab.txt -mapout merged_map.txt\n",
    "\n",
    "###\n",
    "\n",
    "# Will continue with the QC stuff to see if there are other corrections needed before doing the fastx_uniques command again\n",
    "\n",
    "# Examining offset sequences:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -cluster_fast ../merged_otus.fasta -id 0.97 -strand both -alnout otus2.aln -show_termgaps -userout user.txt -userfields query+target+qstrand+qlo+qhi+tlo+thi\n",
    "\n",
    "      Seqs  1301\n",
    "  Clusters  1183\n",
    "  Max size  9\n",
    "  Avg size  1.1\n",
    "  Min size  1\n",
    "Singletons  1149, 88.3% of seqs, 97.1% of clusters\n",
    "   Max mem  707Mb\n",
    "      Time  1.00s\n",
    "Throughput  1301.0 seqs/sec.\n",
    "\n",
    "# Well, all of the qlo values are 1. Not sure what this means...\n",
    "\n",
    "# Looking further into it to compare unique sequences to the OTUs:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -usearch_global ../merged_uniques.fasta -db ../merged_otus.fasta -strand both -id 1.0 -maxaccepts 4 -maxrejects 64 -userout uniques_vs_otus.txt -userfields query+target\n",
    "\n",
    "02:05 735Mb   100.0% Searching merged_uniques.fasta, 44.2% matched\n",
    "\n",
    "cut -f1 uniques_vs_otus.txt | sort | uniq -d > uniq_test_f1.txt\n",
    "cut -f2 uniques_vs_otus.txt | sort | uniq -d > uniq_test_f2.txt \n",
    "\n",
    "# Examine the output files above tomorrow in reference to the following webpage: https://drive5.com/usearch/manual/otu_qc_offset.html\n",
    "\n",
    "## 2020-02-14\n",
    "\n",
    "# Confused because the user.txt output file only shows qlo values of one throughout which means that I don't have \n",
    "\t# staggering or strand sense issues, but then the unique_test_f1.txt shows that many of the uniques have a size of \n",
    "\t# 2. This is conflicting info. Still, think I should go through the pipeline again with the oriented reads for both\n",
    "\t# the merged and R1, unoise and cluster steps...\n",
    "\n",
    "# Instead of dealing with cross talk informatically (i.e. barcode switching), I'll just subtract reads found in the \n",
    "\t# control samples from the rest of the samples, but will need to do it from each pair of months corresponding \n",
    "\t# to the sequencing run. https://drive5.com/usearch/manual/crosstalk.html\n",
    "\n",
    "# Will search for low complexity reads on the final OTUs...\n",
    "\n",
    "https://drive5.com/usearch/manual/cmd_filter_lowc.html\n",
    "\n",
    "# Don't have this command in V10, so will skip it...\n",
    "\n",
    "# Will do a search for phix now:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -search_phix temporal_merged.fastq -notmatched ../OTUs/QC/phix_filtered.fastq -alnout ../OTUs/QC/phix_hits.txt\n",
    "\n",
    "00:00 277Mb   100.0% Word stats\n",
    "00:00 277Mb   100.0% Alloc rows\n",
    "00:00 277Mb   100.0% Build index\n",
    "00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "02:32 936Mb   100.0% Searching, 5.1% matched\n",
    "\n",
    "# There are actually some reads that are coming up as phix in this step, so I'll have to go back before making the fasta file\n",
    "\t# (fastq_filter, I believe) and then do the orientation step, unless these reads don't make it into the final table\n",
    "\t# How to tell if they are actually in the post-processing steps? \n",
    "\t\t# I guess it'll be hard to tell; should go ahead with the removal the phix reads as planned before the fastq_filter and orient command\n",
    "\t\t\n",
    "## Right now, need to do this:\n",
    "- Incorporate the phix_filtered.fastq file (do it for the R1 reads, too!)\n",
    "- Orient the reads\n",
    "\n",
    "## Actually, maybe I need to use the -filter_phix command??? Yep! \n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix temporal_merged.fastq -output phix_filtered.fastq -alnout phix_hits.txt\n",
    "\n",
    "00:00 277Mb   100.0% Word stats\n",
    "00:00 277Mb   100.0% Alloc rows\n",
    "00:00 277Mb   100.0% Build index\n",
    "00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "01:08 933Mb   100.0% Filtering for phix, 838106 hits (5.1%)\n",
    "\n",
    "cp phix_filtered.fastq ../Redux_processing/\n",
    "\n",
    "mv phix_filtered.fastq merged_phix_filtered.fastq\n",
    "\t# Pull the stats from line 859 above\n",
    "\t\n",
    "\t00:00 277Mb   100.0% Word stats\n",
    "\t00:00 277Mb   100.0% Alloc rows\n",
    "\t00:00 277Mb   100.0% Build index\n",
    "\t00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "\t01:08 933Mb   100.0% Filtering for phix, 838106 hits (5.1%)\n",
    "\n",
    "\n",
    "## Remember that there wasn't a merged_cutadapt.fastq file because the merging stripped the primers.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix ../Forward_reads/temporal_R1_cutadapt.fastq -output R1_phix_filtered.fastq -alnout ../Forward_reads/phix_hits.txt\n",
    "\n",
    "00:00 277Mb   100.0% Word stats\n",
    "00:00 277Mb   100.0% Alloc rows\n",
    "00:00 277Mb   100.0% Build index\n",
    "00:00 244Mb  CPU has 28 cores, defaulting to 10 threads\n",
    "01:33 930Mb   100.0% Filtering for phix, 1035882 hits (5.5%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter merged_phix_filtered.fastq -fastq_maxee 1.0 -fastaout merged_filtered.fasta\n",
    "\n",
    "00:01 38Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "01:48 683Mb   100.0% Filtering, 95.6% passed\n",
    "  15486682  Reads (15.5M)                   \n",
    "    677084  Discarded reads with expected errs > 1.00\n",
    "  14809598  Filtered reads (14.8M, 95.6%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter R1_phix_filtered.fastq -fastq_maxee 1.0 -fastaout R1_filtered.fasta\n",
    "\n",
    "00:00 38Mb   CPU has 28 cores, defaulting to 10 threads\n",
    "01:33 683Mb   100.0% Filtering, 85.9% passed\n",
    "  17743132  Reads (17.7M)                   \n",
    "   2502269  Discarded reads with expected errs > 1.00\n",
    "  15240863  Filtered reads (15.2M, 85.9%)\n",
    "  \n",
    "usearch10.0.274M_i86linux64 -orient merged_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout merged_oriented.fasta -tabbedout merged_orient.txt\n",
    "\n",
    "01:41 1.0Gb   100.0% 14678970 plus (99.1%), 985 minus (0.0%), 129643 undet. (0.9%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient R1_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout R1_oriented.fasta -tabbedout R1_orient.txt\n",
    "\n",
    "01:41 1.0Gb   100.0% 15129120 plus (99.3%), 218 minus (0.0%), 111525 undet. (0.7%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques merged_oriented.fasta -fastaout merged_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:45 4.4Gb   100.0% Reading merged_oriented.fasta\n",
    "00:45 4.4Gb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:53 7.8Gb   100.0% DF\n",
    "01:01 7.9Gb  14679955 seqs, 306010 uniques, 199184 singletons (65.1%)\n",
    "01:01 7.9Gb  Min size 1, median 1, max 9957620, avg 47.97\n",
    "01:04 5.6Gb   100.0% Writing merged_uniques.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques R1_oriented.fasta -fastaout R1_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:49 4.6Gb   100.0% Reading R1_oriented.fasta\n",
    "00:49 4.5Gb  CPU has 28 cores, defaulting to 10 threads\n",
    "00:56 8.0Gb   100.0% DF\n",
    "01:04 8.1Gb  15129338 seqs, 683798 uniques, 518290 singletons (75.8%)\n",
    "01:04 8.1Gb  Min size 1, median 1, max 9174463, avg 22.13\n",
    "01:08 5.7Gb   100.0% Writing R1_uniques.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -cluster_otus merged_uniques.fasta -otus merged_otus.fasta -relabel Otu\n",
    "\n",
    "00:13 60Mb    100.0% 1102 OTUs, 65 chimeras\n",
    "\n",
    "usearch10.0.274M_i86linux64 -cluster_otus R1_uniques.fasta -otus R1_otus.fasta -relabel Otu\n",
    "\n",
    "00:19 60Mb    100.0% 1217 OTUs, 104 chimeras\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 merged_uniques.fasta -zotus merged_zotus.fasta\n",
    "\n",
    "00:01 130Mb   100.0% Reading merged_uniques.fasta\n",
    "00:01 120Mb   100.0% 994 amplicons, 2592535 bad (size >= 8)\n",
    "00:15 131Mb   100.0% 987 good, 7 chimeras                  \n",
    "00:15 131Mb   100.0% Writing zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 R1_uniques.fasta -zotus R1_zotus.fasta\n",
    "\n",
    "00:02 229Mb   100.0% Reading R1_uniques.fasta\n",
    "00:03 223Mb   100.0% 987 amplicons, 3535377 bad (size >= 8) \n",
    "00:15 234Mb   100.0% 980 good, 7 chimeras                  \n",
    "00:15 234Mb   100.0% Writing zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab merged_phix_filtered.fastq -otus merged_otus.fasta -otutabout merged_otutab.txt -mapout merged_map.txt\n",
    "\n",
    "42:25 736Mb   100.0% Searching, 99.7% matched\n",
    "15444917 / 15486682 mapped to OTUs (99.7%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab R1_phix_filtered.fastq -otus R1_otus.fasta -otutabout R1_otutab.txt -mapout R1_map.txt\n",
    "\n",
    "43:18 736Mb   100.0% Searching, 96.2% matched\n",
    "17071726 / 17743132 mapped to OTUs (96.2%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab merged_phix_filtered.fastq -otus merged_zotus.fasta -otutabout merged_zotutab.txt -mapout merged_zmap.txt\n",
    "\n",
    "43:58 726Mb   100.0% Searching, 99.6% matched\n",
    "15423973 / 15486682 mapped to OTUs (99.6%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab R1_phix_filtered.fastq -otus R1_zotus.fasta -otutabout R1_zotutab.txt -mapout R1_zmap.txt\n",
    "\n",
    "49:52 726Mb   100.0% Searching, 95.9% matched\n",
    "17018397 / 17743132 mapped to OTUs (95.9%)\n",
    "\n",
    "## 2020-02-19\n",
    "\n",
    "# Okay, now to complete the taxonomy step with the updated udb\n",
    "\t# Should consider using a cutoff of ~80% for the taxonomic assignments, though many of the fungi are pretty unique\n",
    "\t\t# Maybe do an non-cutoff and a cutoff version and see how the results compare? \n",
    "\t\t\t# This would mean that I'd have 8 different analyses to run through. Maybe do a comparison of the \n",
    "\t\t\t# cutoffs from the merged otus for starters?\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Taxonomy\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_otus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout ../Taxonomy/merged_otus.sintax -strand both\n",
    "\n",
    "# Note, no real output from the above command, so nothing printed here\n",
    "\n",
    "# Concerned the applying the 80% cutoff is going to kick out most of the Catenasporaceae OTUs since I see only 2% seq identity for \n",
    "\t# OTU 2 at genus level\n",
    "\t\t# Will try it anyways...\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_otus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout ../Taxonomy/merged_otus_80_percent.sintax -sintax_cutoff 0.8 -strand both\n",
    "\n",
    "# Okay, it looks like it kept OTU2 despite the low similarity. Maybe it only applies the threshold at higher taxonomic ranks? \n",
    "\t# Should consider comparing the two, side-by-side in R, but will proceed with the standard assignment for the remaining\n",
    "\t# otus/zotus for now.\n",
    "\t\n",
    "usearch10.0.274M_i86linux64 -sintax R1_otus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout ../Taxonomy/R1_otus.sintax -strand both\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout ../Taxonomy/merged_zotus.sintax -strand both\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax R1_zotus.fasta -db ~/UNITE2/unite_2020_02_04.udb -tabbedout ../Taxonomy/R1_zotus.sintax -strand both\n",
    "\n",
    "# Ready to move on to R? \n",
    "\t# Will quickly try an octave plot of the merged_otus and the merged_zotus to see how things look. Will then move on to R.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -calc_distmx ../Redux_processing/merged_otus.fasta -tabbedout mx.txt\n",
    "\n",
    "usearch10.0.274M_i86linux64 -calc_distmx ../Redux_processing/merged_zotus.fasta -tabbedout mx_zotus.txt\n",
    "\n",
    "# Nevermind, the otutab_octave command isn't in V10\n",
    "\n",
    "# Okay, will move on to R now...\n",
    "\n",
    "# Well, that's disappointing: only have the apr_may samples in the otu table. Will check to see if this is the case in the zotu tables\n",
    "\n",
    "# The reads are in the merged files, but I suspect the difference in the header lines is throwing off the commands. \n",
    "\t# Recall needing to reformat the header lines for the apr_may samples, but will have to look through the old code to \n",
    "\t# figure out what I did\n",
    "\t\t# I did use a script fastq_header_changer3.py to change the apr_may files previously, but maybe I can get around it with\n",
    "\t\t# the -sample_delim . option for the otu_tab command, if all the reads are present up until that point...\n",
    "\n",
    "fastq_header_changer3.py apr_may.fastq > apr_may_V2.fastq\n",
    "\n",
    "# A quick check of the headers for comparison:\n",
    "\n",
    "less apr_may_V2.fastq: @BY-12-1-5.1101138891730\n",
    "less aug_sep.fastq: @BY-9-4-9.1337858\n",
    "\n",
    "# Looks good\n",
    "\t# Fuck balls. Will need to re-orient and phix_filter these reads again. Should I just use the header_changer3.py command on the\n",
    "\t# phix-filtered file? Definitely...if it works. Will try tomorrow. \n",
    "\t\n",
    "# 2020-02-20\n",
    "\t\n",
    "fastq_header_changer3.py merged_phix_filtered.fastq > merged_phix_filtered_V2.fastq\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered.fastq\n",
    "\t15487871\n",
    "\n",
    "grep -c \"^@\" merged_phix_filtered_V2.fastq \n",
    "\t15487871\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter merged_phix_filtered_V2.fastq -fastq_maxee 1.0 -fastaout merged_filtered_V2.fasta\n",
    "\n",
    "01:44 683Mb   100.0% Filtering, 95.6% passed\n",
    "  15486682  Reads (15.5M)                   \n",
    "    677084  Discarded reads with expected errs > 1.00\n",
    "  14809598  Filtered reads (14.8M, 95.6%)\n",
    "    \n",
    "usearch10.0.274M_i86linux64 -orient merged_filtered_V2.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout merged_oriented_V2.fasta -tabbedout merge_orient_V2.txt\n",
    "\n",
    "01:40 1.0Gb   100.0% 14678970 plus (99.1%), 985 minus (0.0%), 129643 undet. (0.9%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques merged_oriented_V2.fasta -fastaout merged_uniques_V2.fasta -sizeout -relabel Uniq\n",
    "\n",
    "01:02 7.6Gb  14679955 seqs, 306010 uniques, 199184 singletons (65.1%)\n",
    "01:02 7.6Gb  Min size 1, median 1, max 9957620, avg 47.97\n",
    "01:04 5.3Gb   100.0% Writing merged_uniques_V2.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 merged_uniques_V2.fasta -zotus merged_zotus_V2.fasta\n",
    "\n",
    "00:01 130Mb   100.0% Reading merged_uniques_V2.fasta\n",
    "00:02 120Mb   100.0% 994 amplicons, 2592535 bad (size >= 8)  \n",
    "00:16 131Mb   100.0% 987 good, 7 chimeras                  \n",
    "00:16 131Mb   100.0% Writing zotus\n",
    "\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab merged_phix_filtered_V2.fastq -otus merged_zotus_V2.fasta -otutabout merged_zotutab_V3.txt -mapout merged_zmap_V3.txt -sample_delim . \n",
    "\n",
    "42:18 726Mb   100.0% Searching, 99.6% matched\n",
    "15423973 / 15486682 mapped to OTUs (99.6%)\n",
    "42:18 726Mb  Writing merged_zotutab_V3.txt\n",
    "42:18 726Mb  Writing merged_zotutab_V3.txt ...done.\n",
    "\n",
    "\n",
    "/workspace/scratch/obrett/RDP\n",
    "\n",
    "wget https://drive5.com/sintax/rdp_its_v2.fa.gz\n",
    "\n",
    "gunzip rdp_its_v2.fa.gz\n",
    "\n",
    "usearch10.0.274M_i86linux64 -makeudb_sintax rdp_its_v2.fa -output rdp_its_v2.udb\n",
    "\n",
    "## This output has a taxonomic assignment for almost every rank which is helpful. Should compare between the two further. \n",
    "\n",
    "\t# Think I'm going to work with the RDP...\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus_V2.fasta -db ../../../RDP/rdp_its_v2.udb -strand both -tabbedout merged_zotus_V2_rdp.sintax\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Processing for within plant\n",
    "\n",
    "## First need to rename the fastq files (and potentially the headers) to make sure they process neatly downstream\n",
    "\n",
    "## Will modify the shell script jun_jul_renamer.sh for the within-plant\n",
    "\n",
    "within_plant_renamer.sh\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_mergepairs *R1*.fastq -fastqout ../Merged_reads/within_p_merged2.fastq -relabel @ -fastq_maxdiffs 300\n",
    "\n",
    "Totals:\n",
    "    675853  Pairs (675.9k)\n",
    "    583714  Merged (583.7k, 86.37%)\n",
    "    229030  Alignments with zero diffs (33.89%)\n",
    "     74970  Too many diffs (> 300) (11.09%)\n",
    "     17169  No alignment found (2.54%)\n",
    "         0  Alignment too short (< 16) (0.00%)\n",
    "    581026  Staggered pairs (85.97%) merged & trimmed\n",
    "    203.81  Mean alignment length\n",
    "    219.84  Mean merged length\n",
    "      0.37  Mean fwd expected errors\n",
    "      1.73  Mean rev expected errors\n",
    "      0.17  Mean merged expected errors\n",
    "\n",
    "usearch10.0.274M_i86linux64 -filter_phix within_p_merged2.fastq -output within_p_phix_f.fastq -alnout phix_hits.txt\n",
    "\n",
    "00:02 935Mb   100.0% Filtering for phix, 12380 hits (2.1%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastq_filter within_p_phix_f.fastq -fastq_maxee 1.0 -fastaout within_p_filtered.fasta\n",
    "\n",
    "00:04 683Mb   100.0% Filtering, 96.6% passed\n",
    "    571334  Reads (571.3k)                  \n",
    "     19582  Discarded reads with expected errs > 1.00\n",
    "    551752  Filtered reads (551.8k, 96.6%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -orient within_p_filtered.fasta -db ~/UNITE2/unite_2020_02_04.udb -fastaout within_p_oriented.fasta -tabbedout within_p_oriented.txt\n",
    "\n",
    "00:04 1.0Gb   100.0% 548048 plus (99.3%), 2 minus (0.0%), 3702 undet. (0.7%)\n",
    "\n",
    "usearch10.0.274M_i86linux64 -fastx_uniques within_p_oriented.fasta -fastaout within_p_uniques.fasta -sizeout -relabel Uniq\n",
    "\n",
    "00:02 898Mb  548050 seqs, 20995 uniques, 15276 singletons (72.8%)\n",
    "00:02 898Mb  Min size 1, median 1, max 180839, avg 26.10\n",
    "00:03 898Mb   100.0% Writing within_p_uniques.fasta\n",
    "\n",
    "usearch10.0.274M_i86linux64 -unoise3 within_p_uniques.fasta -zotus within_p_zotus.fasta\n",
    "\n",
    "00:00 49Mb    100.0% Reading within_p_uniques.fasta\n",
    "00:00 30Mb    100.0% 620 amplicons, 38994 bad (size >= 8)\n",
    "00:05 41Mb    100.0% 615 good, 5 chimeras                \n",
    "00:05 41Mb    100.0% Writing zotus\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab within_p_phix_f.fastq -otus within_p_zotus.fasta -otutabout within_p_zotutab.txt -mapout within_p_zmap.txt\n",
    "\n",
    "03:32 726Mb   100.0% Searching, 99.1% matchedchedg, 98.6% matched\n",
    "566207 / 571334 mapped to OTUs (99.1%)\n",
    "03:32 726Mb  Writing within_p_zotutab.txt\n",
    "03:32 726Mb  Writing within_p_zotutab.txt ...done.\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax within_p_zotus.fasta -db ../../../RDP/rdp_its_v2.udb -tabbedout within_p_zotus.sintax -strand both\n",
    "\n",
    "## Done. Will incorporate into R for future analyses\n",
    "\n",
    "# Just copied over the following files to the following directory for the # # within plant analysis:\n",
    "\n",
    "/Users/brett/Documents/Temporal_turnover/Informatics\n",
    "within_p_zotus.sintax\n",
    "within_p_zotutab.txt\n",
    "\n",
    "# Need to make an updated RDP sintax file for the zotu table:\n",
    "\n",
    "usearch10.0.274M_i86linux64 -sintax merged_zotus_V2.fasta -db ../../../RDP/rdp_its_v2.udb -strand both -tabbedout merged_zotus_V2_rdp.sintax\n",
    "\n",
    "# Need the following files for zotu analysis:\n",
    "- merged_zotus_V2.fasta\n",
    "- merged_zotus_V2_rdp.sintax\n",
    "- merged_zotutab_V3.txt\n",
    "- map_file16.txt\n",
    "\n",
    "# Need to reformat merged_zotus_V2_rdp.sintax so it matches the formatting of merged_otus_V3.sintax\n",
    "\t# Saved as merged_zotus_V3_rdp.sintax\n",
    "\n",
    "# Consider removing conf estimates and incertae sedis from tax table\n",
    "\n",
    "# Need to re-run line 1629 from above with the -sample_delim . argument\n",
    "\n",
    "/workspace/scratch/obrett/Temporal/Within_plant/Merged_reads\n",
    "\n",
    "usearch10.0.274M_i86linux64 -otutab within_p_phix_f.fastq -otus within_p_zotus.fasta -otutabout within_p_zotutab2.txt -mapout within_p_zmap2.txt -sample_delim .\n",
    "\n",
    "within_p_zotus.sintax\n",
    "within_p_zotutab2.txt\n",
    "within_p_map_file.txt\n",
    "\n",
    "# Plotted an ordination and taxonomy plot for the within-p\n",
    "\t# See that there are multiples of the same genus/family on the \n",
    "\t# taxonomy plot. Need to delete these conf. estimates since it makes \n",
    "\t# each genus seem distinct. Will also delete the f: and g: designations\n",
    "\t\n",
    "# Will update this on both taxonomy files for the big dataset and the within-p\n",
    "\n",
    "Originals:\n",
    "- 'within_p_zotus2.sintax.txt'\n",
    "- 'merged_zotus_V3_rdp.sintax'\n",
    "\n",
    "Edited:\n",
    "- 'within_p_zotus3.sintax'\n",
    "- 'merged_zotus_V4_rdp.sintax'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
